rcp/org.testeditor.rcp4.views.tcltestrun/src/org/testeditor/rcp4/views/tcltestrun/TclLauncherUi.xtend

```diff
@@ -98,122 +98,98 @@ class TclLauncherUi implements Launcher {
 		return result.get
 	}
 
-	private def boolean launchTest(IStructuredSelection selection, IProject project, String elementId,
-		TclLauncher launcher, Map<String, Object> options) {
+	private def boolean launchTest(List<String> testCasesCommaList, IProject project, TclLauncher launcher,
+		Map<String, Object> options) {
 		logger.info("Trying to launch launcherClass='{}' test execution for elementId='{}' in project='{}'",
-			launcher.class.simpleName, elementId, project)
+			launcher.class.simpleName, testCasesCommaList.get(0), project)
 		progressRunner.run([ monitor |
-			monitor.beginTask("Test execution: " + elementId, IProgressMonitor.UNKNOWN)
-			val testCasesCommaList = createTestCasesCommaList(selection)
+			monitor.beginTask("Test execution: " + testCasesCommaList.get(0), IProgressMonitor.UNKNOWN)
+
 			val result = launcher.launchTest(testCasesCommaList, project, monitor, options)
 			project.refreshLocal(IProject.DEPTH_INFINITE, monitor)
 			if (result.expectedFileRoot == null) {
 				logger.error("resulting expectedFile must not be null")
 			} else {
-				safeUpdateJunitTestView(elementId, result.expectedFileRoot, project.name)
+				safeUpdateJunitTestView(result.expectedFileRoot, project.name)
 			}
 			monitor.done
 		])
 		return true
 	}
 
-	def String createTestCasesCommaList(IStructuredSelection selection) {
-		if (launchShortcutUtil == null) {
-			launchShortcutUtil = tclInjectorProvider.get.getInstance(LaunchShortcutUtil)
+	def List<String> createGradleTestCasesList(IStructuredSelection selection) {
+		val result = new ArrayList<String>()
+		for (element : selection.toList) {
```

Comment:
The refactored code doesn't look better, I think.
What missed I?

		result += selection.toList.filter(IResource).map [ resource |
			if (resource instanceof IFolder) {
				val javaElement = resource.getAdapter(IJavaElement) as IJavaElement
				if (javaElement != null) {
					return (javaElement.elementName + "*")
				} else {
					logger.warn("selected element {} is not a test exeutuable", resource.name)
					return null
				}

			} else {
				return launchShortcutUtil.getQualifiedNameForTestInTcl(resource).toString
			}
		].filterNull